### PR TITLE
add interfaces to MachineManagementMixins for shutdown/startup/reboot…

### PR DESCRIPTION
**Pull request by ahgittin**
… and for getting metrics

implementation using jclouds does suspend, but as a placeholder for other providers to do true stops/starts

### CODE REVIEW

File: locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java [hunk 1]
```diff
@@ -20,6 +20,7 @@
 
 import static com.google.common.base.Preconditions.checkArgument;
 import static com.google.common.base.Preconditions.checkNotNull;
+import org.apache.brooklyn.api.location.MachineManagementMixins.MachineMetadata;
```
Comment (grkvlt) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java:23]: Reorg imports?

File: locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java [hunk 2]
```diff
@@ -1256,6 +1257,39 @@ public JcloudsMachineLocation resumeMachine(Map<?, ?> flags) {
...
+        if (!(l instanceof JcloudsSshMachineLocation)) {
+            throw new IllegalStateException("Cannot startup machine "+l+"; wrong type");
+        }
+        getComputeService().rebootNode(((JcloudsSshMachineLocation)l).getJcloudsId());
+        return l;
+    }
+
+    @Override
+    public Map<String, Object> getMachineMetrics(MachineLocation l) {
+        return MutableMap.of();
```
Comment (grkvlt) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java:1285]: Why mutable? I would assume there is no expectation the recipient can modify this?

**Review by grkvlt — CHANGES_REQUESTED** [review]
LGTM modulo two comments

File: locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java (hunk 2 above)
Comment (ahgittin) [reply locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/JcloudsLocation.java:1320]: a subclass might

**Comment by ahgittin** [comment]
some minor adjustments to make it better handle management/tracking of locations across suspend/resume and shutdown/restart cycles

@grkvlt thx for the early PR comments, addressed/responded

File: core/src/main/java/org/apache/brooklyn/core/location/MachineLifecycleUtils.java [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+                String m1id = ((GivesMachineMetadata)p1).getMachineMetadata(m1).getId();
+                String m2id = ((GivesMachineMetadata)p2).getMachineMetadata(m2).getId();
+                if (Objects.equals(m1id, m2id)) {
+                    if (m1id!=null) {
+                        primarilySame = true;
+                    }  else {
+                        // indeterminate
+                    }
+                } else {
+                    primarilySame = false;
```
Comment (duncangrant) [inline core/src/main/java/org/apache/brooklyn/core/location/MachineLifecycleUtils.java:261]: Could return false here?

**Comment by duncangrant** [comment]
@ahgittin LGTM - I've created a PR to fix https://github.com/apache/brooklyn-library/pull/182 - obviously close this if you don't want it

**Review by duncangrant — APPROVED** [review]

**Pull request merged**